flask/cli.py

```diff
@@ -520,15 +520,37 @@ def load_dotenv():
 
     .. _python-dotenv: https://github.com/theskumar/python-dotenv#readme
 
+    :param names: List of file names to search for.
+    :param top_dir: Don't load files above this directory. If the current
+        directory is above this directory, it is set to this directory before
+        searching, then reset afterwards. This must be an absolute path.
+    :param chdir: Whether to set the current directory to the location of the
+        first file found.
+
     .. versionadded:: 1.0
     """
 
     if dotenv is None:
         return
 
+    old_dir = None
     new_dir = None
 
-    for name in ('.env', '.flaskenv'):
+    if top_dir is not None:
+        if not os.path.isabs(top_dir):
+            raise ValueError("'top_dir' must be an absolute path.")
+
+        cwd = os.path.realpath(os.getcwd())
+        top_dir = os.path.realpath(top_dir)
+
+        # If the top dir is below the current dir, store the current dir and
+        # change to the top dir temporarily. This disables setting the dir
+        # based on found env files.
+        if os.path.commonprefix((cwd, top_dir)) == cwd:
```

Comment:
What happens if `top_dir` is not below `cwd`? If the reverse is true (`cwd` below `top_dir`) then you can leave the cwd alone and let dotenv do its lookup, but if `cwd` and `top_dir` are in different branches, I think you still want to change the cwd to `top_dir`. This would be the use case of having an unrelated cwd, and still wanting to import your settings from the .env file located with the app.